titolo
P01 = dataset associazione CIG-CUP
P03 = dataset CIG-misure premiali
P04 = dataset CIG-deroghe 
P06 = dataset CIG-CUP-anagrafica
Tentativi di merge di P06 con P03/P04

In [105]:
# dependencies
import pandas as pd
import numpy as np
pd.set_option('display.float_format', lambda x: '%.9f' % x)
pd.set_option('display.float_format', str)

In [106]:
# datasets
p01 = pd.read_csv("./datasets/cup_csv.csv",delimiter=';')
p03 = pd.read_excel("./datasets/p06-p03-p04_formattati.xlsx", sheet_name="misurepremiali-pnrrpnc_csv")
# dtype=str
p04 = pd.read_excel("./datasets/p06-p03-p04_formattati.xlsx", sheet_name="indicatori-pnrrpnc_csv")
p06 = pd.read_excel("./datasets/p06-p03-p04_formattati.xlsx", sheet_name="PNRR_Gare-Dati_Validati")

In [125]:
p04 = pd.read_excel("./datasets/p06-p03-p04_formattati.xlsx", sheet_name="indicatori-pnrrpnc_csv",dtype=str)

Operazioni sui dataset di preprocessing per poter fare merge

In [126]:
p01 = p01[["CIG", "CUP"]]
p03 = p03.rename(columns={'cig':'CIG'})
p04 = p04.rename(columns={'cig':'CIG'})

In [127]:
p04

CIG flag_quote quota_femminile quota_giovanile cod_mot_deroga  \
0       999337491A          N             NaN             NaN              1   
1       947945126F          N             NaN             NaN              8   
2       9494177AB4          N             NaN             NaN              1   
3       9475802F2B          N             NaN             NaN              1   
4       9475802F2B          N             NaN             NaN              8   
...            ...        ...             ...             ...            ...   
193045  A020B19D50          S             NaN             NaN            NaN   
193046  A0035B63AE          S             NaN             NaN            NaN   
193047  A01C6CC040        NaN             NaN             NaN            NaN   
193048  A01C41B7A9          S             NaN             NaN            NaN   
193049  A011D51429        NaN             NaN             NaN            NaN   

                           mot_deroga  
0       IMPORTO RIDOTTO DEL CONTRATTO  
1                               ALTRO  
2       IMPORTO RIDOTTO DEL CONTRATTO  
3       IMPORTO RIDOTTO DEL CONTRATTO  
4                               ALTRO  
...                               ...  
193045                            NaN  
193046                            NaN  
193047                            NaN  
193048                            NaN  
193049                            NaN  

[193050 rows x 6 columns]

In [120]:
# per ri aggiustare i type
p04 = p04.astype({'quota_femminile': float,'quota_giovanile':float})
# p04['cod_mot_deroga'] = p04['cod_mot_deroga'].fillna(0).astype(int)
# p04['cod_mot_deroga'].replace(0, np.nan,inplace=True)


In [128]:
p06.head(5)
print('shape p01:',p01.shape[0])
print('shape p03:',p03.shape[0])
print('shape p04:',p04.shape[0])
print('shape p06:',p06.shape[0])

shape p01: 1048575
shape p03: 25278
shape p04: 193050
shape p06: 65144


In [153]:
# merging p06/p04 on CIG
d_46 = p06.merge(p04,on="CIG",how='inner')
print('shape of p06/p04 merged:', d_46.shape[0])
#d_46 = d_46.drop_duplicates()
#print('shape of p06/p04 merged:', d_46.shape[0])

shape of p06/p04 merged: 23596


In [154]:
# merging p04/p01 on CIG
d_14 = p04.merge(p01,on='CIG',how='inner')
print('shape of p04/p01 merged:',d_14.shape[0])
#d_14 = d_14.drop_duplicates()
#print('shape of p04/p01 merged:', d_14.shape[0])

shape of p04/p01 merged: 56798


Non tutti i CIG di p04 (deroghe) ci sono in p06. I matching CIG sono 23596, a fronte di 193050 CIG in P04. Progetti non andati a buon fine? Alcuni CIG di p04 che non trovano corrispondenza in p06 la trovavano in p01, ma non tutti (i matching cig tra p04 e p01 sono 56798). 


In [155]:
# merging p04/p03 on CIG
d_34 = p04.merge(p03,on='CIG',how='inner')
print('shape of p04/p03 merged:',d_34.shape[0])
#d_34 = d_34.drop_duplicates()
#print('shape of p04/p03 merged:',d_34.shape[0])
#print(d_34['CIG'].nunique())

shape of p04/p03 merged: 28349


In [156]:
# merging p06/p03 on CIG
d_36 = p06.merge(p03,on="CIG",how='inner')
print('shape temp p06-p03:',d_36.shape[0])
#d_36 = d_36.drop_duplicates()
#print('shape temp p06-p03:',d_36.shape[0])

shape temp p06-p03: 5495


In [152]:
# merging p06/p01 on CIG
d_16 = p06.merge(p01,on="CIG",how='inner')
print('shape temp p06-p01:',d_16.shape[0])

shape temp p06-p01: 24212
